daml-script/dump/src/main/scala/com/daml/script/dump/TreeUtils.scala

```diff
@@ -159,39 +160,97 @@ object TreeUtils {
   def createdReferencedCids(ev: CreatedEvent): Set[ContractId] =
     ev.createArguments.foldMap(args => args.fields.foldMap(f => valueCids(f.getValue.sum)))
 
-  /** A simple event causes the creation of a single contract and returns its contract id.
+  def cmdReferencedCids(cmd: Command): Set[ContractId] = {
+    cmd match {
+      case CreateCommand(createdEvent) =>
+        createdReferencedCids(createdEvent)
+      case ExerciseCommand(exercisedEvent) =>
+        exercisedEvent.choiceArgument.foldMap(arg => valueCids(arg.sum)) + ContractId(
+          exercisedEvent.contractId
+        )
+      case CreateAndExerciseCommand(createdEvent, exercisedEvent) =>
+        createdReferencedCids(createdEvent) ++ exercisedEvent.choiceArgument.foldMap(arg =>
+          valueCids(arg.sum)
+        )
+    }
+  }
+
+  sealed trait Command
+  final case class CreateCommand(createdEvent: CreatedEvent) extends Command
+  final case class ExerciseCommand(exercisedEvent: ExercisedEvent) extends Command
+  final case class CreateAndExerciseCommand(
+      createdEvent: CreatedEvent,
+      exercisedEvent: ExercisedEvent,
+  ) extends Command
+
+  object Command {
+    def fromTree(tree: TransactionTree): Seq[Command] = {
+      val commands = mutable.LinkedHashMap.empty[EventId, Command]
+      val created = mutable.HashMap.empty[ContractId, EventId]
+      tree.rootEventIds.foreach { eventId =>
+        val event = tree.eventsById(eventId)
+        event.kind match {
+          case Kind.Empty =>
+          case Kind.Created(createdEvent) =>
+            commands += EventId(eventId) -> CreateCommand(createdEvent)
+            created += ContractId(createdEvent.contractId) -> EventId(eventId)
+          case Kind.Exercised(exercisedEvent) =>
+            created.get(ContractId(exercisedEvent.contractId)) match {
+              case None =>
+                commands += EventId(eventId) -> ExerciseCommand(exercisedEvent)
+              case Some(createdEventId) =>
+                commands.get(createdEventId) match {
+                  case Some(CreateCommand(createdEvent)) =>
+                    // Replace the original CreateCommand
+                    commands += createdEventId -> CreateAndExerciseCommand(
+                      createdEvent,
+                      exercisedEvent,
+                    )
+                  case _ =>
+                    throw new RuntimeException(
+                      "Internal error: Encountered `created` entry without corresponding `commands` entry."
+                    )
+                }
+            }
+        }
+      }
+      commands.valuesIterator.toSeq
```

Comment:
This seems to return a more or less random order. I think you have to keep track of the order in which you encounter events (I don’t think event ids provide ordering guarantees).

Reply:
`commands` is a `LinkedHashMap` which returns items in insertion order. So the order is not random, but the order in which we encounter events.